sycl/include/CL/sycl/detail/spirv.hpp

```diff
@@ -41,47 +67,137 @@ template <typename Group> bool GroupAny(bool pred) {
+
 // Broadcast with scalar local index
 // Work-group supports any integral type
 // Sub-group currently supports only uint32_t
+template <typename Group> struct GroupId { using type = size_t; };
+template <> struct GroupId<::cl::sycl::intel::sub_group> {
+  using type = uint32_t;
+};
 template <typename Group, typename T, typename IdT>
-detail::enable_if_t<is_group<Group>::value && std::is_integral<IdT>::value, T>
-GroupBroadcast(T x, IdT local_id) {
+EnableIfNativeBroadcast<T, IdT> GroupBroadcast(T x, IdT local_id) {
+  using GroupIdT = typename GroupId<Group>::type;
+  GroupIdT GroupLocalId = static_cast<GroupIdT>(local_id);
   using OCLT = detail::ConvertToOpenCLType_t<T>;
-  using OCLIdT = detail::ConvertToOpenCLType_t<IdT>;
-  OCLT ocl_x = detail::convertDataToType<T, OCLT>(x);
-  OCLIdT ocl_id = detail::convertDataToType<IdT, OCLIdT>(local_id);
-  return __spirv_GroupBroadcast(group_scope<Group>::value, ocl_x, ocl_id);
+  using OCLIdT = detail::ConvertToOpenCLType_t<GroupIdT>;
+  OCLT OCLX = detail::convertDataToType<T, OCLT>(x);
+  OCLIdT OCLId = detail::convertDataToType<GroupIdT, OCLIdT>(GroupLocalId);
+  return __spirv_GroupBroadcast(group_scope<Group>::value, OCLX, OCLId);
 }
 template <typename Group, typename T, typename IdT>
-detail::enable_if_t<is_sub_group<Group>::value && std::is_integral<IdT>::value,
-                    T>
-GroupBroadcast(T x, IdT local_id) {
-  using SGIdT = uint32_t;
-  SGIdT sg_local_id = static_cast<SGIdT>(local_id);
-  using OCLT = detail::ConvertToOpenCLType_t<T>;
-  using OCLIdT = detail::ConvertToOpenCLType_t<SGIdT>;
-  OCLT ocl_x = detail::convertDataToType<T, OCLT>(x);
-  OCLIdT ocl_id = detail::convertDataToType<SGIdT, OCLIdT>(sg_local_id);
-  return __spirv_GroupBroadcast(group_scope<Group>::value, ocl_x, ocl_id);
+EnableIfBitcastBroadcast<T, IdT> GroupBroadcast(T x, IdT local_id) {
+  using GroupIdT = typename GroupId<Group>::type;
+  GroupIdT GroupLocalId = static_cast<GroupIdT>(local_id);
+  using BroadcastT = ConvertToNativeBroadcastType_t<T>;
+  using OCLIdT = detail::ConvertToOpenCLType_t<GroupIdT>;
+  auto BroadcastX = detail::bit_cast<BroadcastT>(x);
+  OCLIdT OCLId = detail::convertDataToType<GroupIdT, OCLIdT>(GroupLocalId);
+  BroadcastT Result =
+      __spirv_GroupBroadcast(group_scope<Group>::value, BroadcastX, OCLId);
+  return detail::bit_cast<T>(Result);
+}
+template <typename Group, typename T, typename IdT>
+EnableIfGenericBroadcast<T, IdT> GroupBroadcast(T x, IdT local_id) {
+  T Result;
+  char *XBytes = reinterpret_cast<char *>(&x);
+  char *ResultBytes = reinterpret_cast<char *>(&Result);
+  auto BroadcastBytes = [=](size_t Offset, size_t Size) {
+    uint64_t BroadcastX, BroadcastResult;
+    detail::memcpy(&BroadcastX, XBytes + Offset, Size);
+    BroadcastResult = GroupBroadcast<Group>(BroadcastX, local_id);
+    detail::memcpy(ResultBytes + Offset, &BroadcastResult, Size);
+  };
+  GenericShuffle<T>(BroadcastBytes);
```

Comment:
It looks a bit strange to see a broadcast implemented via shuffle. Probably it makes sense to rename `GenericShuffle` to something like "ForEach..something?". I can't come up with a name right now, probably you have any ideas? I wanted to express the idea that functor is applied for each "native" chunk of data in `T`

Reply:
I agree.  This is one of the things that I lazily lumped under the grand refactoring TODO. :)

I have a suspicion that something like this function is going to be useful for implementing other generic versions of SPIR-V functions, but I'm also struggling to come up with a better way to represent it.

I've implemented something in f0a1e2f2a697dadc54bb0539cd2cfa3baad1072b.  It's still not perfect, but is hopefully clearer than what was there before, and should help anybody trying to make sense of the code before these header files get tidied up.